### PR TITLE
Move Generics from MethodSig to TraitItem and ImplItem

### PR DESCRIPTION
**Pull request by sunjay**
As part of `rust-impl-period/WG-compiler-traits`, we want to "lift" `Generics` from `MethodSig` into `TraitItem` and `ImplItem`. This is in preparation for adding associated type generics. (https://github.com/rust-lang/rust/issues/44265#issuecomment-331172238)

Currently this change is only made in the AST. In the future, it may also impact the HIR. (Still discussing)

To understand this PR, it's probably best to start from the changes to `ast.rs` and then work your way to the other files to understand the far reaching effects of this change.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/hir/lowering.rs
Comment (sunjay) [inline src/librustc/hir/lowering.rs]: We should consider updating the HIR to match AST if that is desireable. I have made everything fit for now, but there is now an incongruencey between the structure of the HIR and AST. That's what forced me to add this extra paramater for generics all throughout the code.

File: src/librustc_resolve/lib.rs
Comment (sunjay) [inline src/librustc_resolve/lib.rs]: I'm worried about whether this change has any side-effects we don't want. I don't think I found a place to add this back in after I deleted it.

File: src/librustc_resolve/lib.rs
Comment (sunjay) [inline src/librustc_resolve/lib.rs]: Does this change have any unintended sife-effects?

File: src/libsyntax/parse/parser.rs
Comment (sunjay) [inline src/libsyntax/parse/parser.rs]: Is using `::default()` the right thing to do here?

File: src/libsyntax/parse/parser.rs
Comment (sunjay) [inline src/libsyntax/parse/parser.rs]: I had to make this return type even larger since Generics is outside MethodSig now. There was a lot of code relying on that structure.

File: src/libsyntax/visit.rs
Comment (sunjay) [inline src/libsyntax/visit.rs]: I had to guess here and some other places that this was the right place to add this code. Would be great if you could check.

**Comment by sunjay** [comment]
@nikomatsakis I am currently experiencing a compiler panic when I try to compile this. Given that I am only compiling the compiler and not actually running it, I don't understand where I could have introduced the bug. I would love to know what's going on and would appreciate it if you could give me an idea of how to troubleshoot. The error message I'm receiving is below. Once this is fixed I would like to run the tests and fix any other errors that are found before this gets merged.

```
$ RUST_BACKTRACE=1 ./x.py build
Updating submodules
    Finished dev [unoptimized] target(s) in 0.0 secs
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.0 secs
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage0 test artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.0 secs
Copying stage0 test from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage0 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling syntax v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/libsyntax)
   Compiling rustc_const_math v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_const_math)
   Compiling proc_macro v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/libproc_macro)
   Compiling rustc_back v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_back)
   Compiling syntax_ext v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/libsyntax_ext)
   Compiling rustc v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc)
   Compiling rustc_allocator v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_allocator)
   Compiling rustc_incremental v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_incremental)
   Compiling rustc_const_eval v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_const_eval)
   Compiling rustc_resolve v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_resolve)
   Compiling rustc_metadata v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_metadata)
   Compiling rustc_trans_utils v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_trans_utils)
   Compiling rustc_typeck v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_typeck)
   Compiling rustc_privacy v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_privacy)
   Compiling rustc_passes v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_passes)
   Compiling rustc_mir v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_mir)
   Compiling rustc_lint v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_lint)
   Compiling rustc_trans v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_trans)
   Compiling rustc_save_analysis v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_save_analysis)
   Compiling rustc_borrowck v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_borrowck)
   Compiling rustc_plugin v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_plugin)
   Compiling rustc_driver v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_driver)
   Compiling rustc-main v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/rustc)
    Finished release [optimized] target(s) in 718.57 secs
Copying stage0 rustc from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Assembling stage1 compiler (x86_64-unknown-linux-gnu)
Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling core v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/libcore)
   Compiling unwind v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/libunwind)
   Compiling libc v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/rustc/libc_shim)
   Compiling compiler_builtins v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/rustc/compiler_builtins_shim)
   Compiling rustc_tsan v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_tsan)
   Compiling rustc_lsan v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_lsan)
   Compiling rustc_asan v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_asan)
   Compiling rustc_msan v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/librustc_msan)
   Compiling alloc_jemalloc v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/liballoc_jemalloc)
   Compiling std v0.0.0 (file:///home/sunjay/Documents/projects/rust/src/libstd)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.22.0-dev running on x86_64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'path resolved multiple times (PathResolution { base_def: Trait(DefId { krate: CrateNum(0), node: DefIndex(0:3185) }), unresolved_segments: 0 } before, PathResolution { base_def: Trait(DefId { krate: CrateNum(0), node: DefIndex(0:3185) }), unresolved_segments: 0 } now)', src/librustc_resolve/lib.rs:3460:12
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::_print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: std::panicking::begin_panic_fmt
   7: rustc_resolve::Resolver::record_def
   8: rustc_resolve::Resolver::smart_resolve_path_fragment
   9: rustc_resolve::Resolver::smart_resolve_path
  10: <rustc_resolve::Resolver<'a> as syntax::visit::Visitor<'tcx>>::visit_generics
  11: syntax::visit::walk_impl_item
  12: rustc_resolve::Resolver::with_type_parameter_rib
  13: rustc_resolve::Resolver::with_current_self_type
  14: rustc_resolve::Resolver::with_self_rib
  15: rustc_resolve::Resolver::with_optional_trait_ref
  16: rustc_resolve::Resolver::with_self_rib
  17: rustc_resolve::Resolver::resolve_item
  18: syntax::visit::walk_item
  19: rustc_resolve::Resolver::resolve_item
  20: rustc_resolve::Resolver::resolve_crate
  21: rustc_driver::driver::phase_2_configure_and_expand
  22: rustc_driver::driver::compile_input
  23: rustc_driver::run_compiler

error: Could not compile `core`.

Caused by:
  process didn't exit successfully: `/home/sunjay/Documents/projects/rust/build/bootstrap/debug/rustc --crate-name core src/libcore/lib.rs --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=3181dd9e46400ebd -C extra-filename=-3181dd9e46400ebd --out-dir /home/sunjay/Documents/projects/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/home/sunjay/Documents/projects/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/home/sunjay/Documents/projects/rust/build/x86_64-unknown-linux-gnu/stage1-std/release/deps` (exit code: 101)
warning: build failed, waiting for other jobs to finish...
error: build failed
thread 'main' panicked at 'command did not execute successfully: "/home/sunjay/Documents/projects/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/home/sunjay/Documents/projects/rust/src/libstd/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101', src/bootstrap/compile.rs:883:8
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:611
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:572
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: bootstrap::compile::run_cargo
   8: <bootstrap::compile::Std as bootstrap::builder::Step>::run
   9: bootstrap::builder::Builder::ensure
  10: <bootstrap::compile::Test as bootstrap::builder::Step>::run
  11: bootstrap::builder::Builder::ensure
  12: <bootstrap::compile::Rustc as bootstrap::builder::Step>::run
  13: bootstrap::builder::Builder::ensure
  14: <bootstrap::compile::Assemble as bootstrap::builder::Step>::run
  15: bootstrap::builder::Builder::ensure
  16: bootstrap::builder::Builder::compiler
  17: <bootstrap::compile::Std as bootstrap::builder::Step>::make_run
  18: bootstrap::builder::StepDescription::maybe_run
  19: bootstrap::builder::StepDescription::run
  20: bootstrap::builder::Builder::run
  21: bootstrap::Build::build
  22: bootstrap::main
  23: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  24: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:61
  25: main
  26: __libc_start_main
  27: _start
failed to run: /home/sunjay/Documents/projects/rust/build/bootstrap/debug/bootstrap build
Build completed unsuccessfully in 0:12:04
```

File: src/librustc_resolve/lib.rs
Comment (petrochenkov) [reply src/librustc_resolve/lib.rs]: @sunjay 
You have to visit generics inside of `with_type_parameter_rib` here and for trait items.
This was done previously for methods in this piece of code below:
```rust
this.with_type_parameter_rib(type_parameters, |this| {
    visit::walk_impl_item(this, impl_item); // We visited method signature, `FnKind::Method` and its generics here previously
});
```

File: src/librustc_resolve/lib.rs
Comment (nikomatsakis) [reply src/librustc_resolve/lib.rs]: Indeed, we will need to add this back somewhere =) I think the problem is that the `FnKind` enum has to change. Methods should operate more analogously with `ItemFn` -- so we should add a `&'a Generics` to [the `Method` variant](https://github.com/rust-lang/rust/blob/0962b8fe4ffc1face685bc1b04364f6a3f0ad1ea/src/libsyntax/visit.rs#L37), and then when we construct `FnKind::Method` ([here](https://github.com/rust-lang/rust/blob/0962b8fe4ffc1face685bc1b04364f6a3f0ad1ea/src/libsyntax/visit.rs#L571) and [here](https://github.com/rust-lang/rust/blob/0962b8fe4ffc1face685bc1b04364f6a3f0ad1ea/src/libsyntax/visit.rs#L594)) we can add the data from the trait or impl item respectively. Then we can restore the call to `visit_generics` that occurs *here*, I suppose.

Comment (nikomatsakis) [reply src/librustc_resolve/lib.rs]: Well, let me read a bit more into this actually. I think what I said is not wrong but a few more tweaks are likely needed.

File: src/librustc_resolve/lib.rs
Comment (nikomatsakis) [reply src/librustc_resolve/lib.rs]: Yes, this change is not quite right. =) 

The way that resolve works, it has these "ribs" that are in scope, those are the sets of names that can be used. When you *visit* the generics, you are then resolving the names that appear within. So when you visit the generics here, we are visiting without any names in scope.

What we want to do then is to move up the code that is specific to methods below, and execute it for *all* impl-items:

```rust
                                         let type_parameters =
                                            HasTypeParameters(&impl_item.generics,
                                                             MethodRibKind(!sig.decl.has_self()));
                                         this.with_type_parameter_rib(type_parameters, |this| {...}
```

in that case, we can `visit_generics` here, and you can .. probably forget what I said about modifying  `FnKind::Method` above.

File: src/libsyntax/parse/parser.rs
Comment (nikomatsakis) [reply src/libsyntax/parse/parser.rs]: uh... I'm not sure. =) I wasn't aware that it implemented `default`. But it's probably right.

File: src/libsyntax/parse/parser.rs
Comment (nikomatsakis) [reply src/libsyntax/parse/parser.rs]: yep that is expected

File: src/libsyntax/visit.rs
Comment (nikomatsakis) [reply src/libsyntax/visit.rs]: this seems correct, yes.

File: src/librustc_resolve/lib.rs
Comment (nikomatsakis) [reply src/librustc_resolve/lib.rs]: Actually, I don't think we want to call `visit_generics` here. It will happen as part of the `walk_impl_item` that we see below.

File: src/libsyntax/visit.rs
Comment (nikomatsakis) [inline src/libsyntax/visit.rs]: ...i.e., here, when resolve invokes `walk_impl_item`, it will walk the generics here

File: src/librustc_resolve/lib.rs
Comment (nikomatsakis) [reply src/librustc_resolve/lib.rs]: Yeah, so I think it's fine to remove the call here -- it is being *moved* into the `walk_trait_item` and `walk_impl_item` code, effectively.

File: src/librustc_resolve/lib.rs
Comment (nikomatsakis) [inline src/librustc_resolve/lib.rs]: we should not be invoking `visit_generics` here. IT will happen when we invoke `walk_trait_item` below. Instead, we should be inserting the type-parameters rib:

```rust
let type_parameters = HasTypeParameters(&trait_item.generics, MethodRibKind(!sig.decl.has_self()));
this.with_type_parameter_rib(type_parameters, |this| { /* old code goes here */ })
```

File: src/librustc/hir/lowering.rs
Comment (nikomatsakis) [reply src/librustc/hir/lowering.rs]: I think we will want to update the HIR. I had originally thought we might do that in a separate PR, but I'm not opposed to doing it in the same PR either. It *should* though be possible to get this code to work without doing it, might be easier for testing and things to start that way (then maybe do the HIR change in a follow-up commit).

Comment (sunjay) [reply src/librustc/hir/lowering.rs]: I think it'll be easier to the the HIR in a follow-up PR. I'll get started on that as soon as this is ready to be merged.

File: src/librustc_resolve/lib.rs
Comment (sunjay) [reply src/librustc_resolve/lib.rs]: @nikomatsakis I don't think I can actually move the `HasTypeParameters` code above because it uses `sig` which is only in scope for methods. After this refactoring, I may have to come back and add `HasTypeParameters` for `TraitItemKind::Type` in this match statement.

File: src/librustc_resolve/lib.rs
Comment (sunjay) [reply src/librustc_resolve/lib.rs]: Is this change really okay? If `ItemFn` visits generics, shouldn't `Method` visit them too?

File: src/librustc_resolve/lib.rs
Comment (sunjay) [inline src/librustc_resolve/lib.rs]: I think we'll need to add this type parameter rib to each variant in this match statement because each `HasTypeParamters` will need a different rib kind. Since nothing other than methods has generics right now, would it work to save that change until we are actually implementing associated type generics?

(this applies to both trait items and impl items)

File: src/librustc_resolve/lib.rs
Comment (nikomatsakis) [reply src/librustc_resolve/lib.rs]: Nope. It happens as part of `visit_trait_item` or `visit_impl_item`, I think.

File: src/librustc_resolve/lib.rs
Comment (nikomatsakis) [reply src/librustc_resolve/lib.rs]: Well, it depends. I would think we just want to rename `MethodRibKind` to something else like `TraitOrImplItemRibKind`, but it might be important to distinguish those cases for error messages? I kinda' doubt it, but have to look around.

**Review by nikomatsakis — APPROVED** [review]
At first glance, this is looking good!

**Comment by sunjay** [comment]
@nikomatsakis The build passed! :tada: :tada: :tada: :tada: :tada:

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
OK, so this failed building the RLS. I'm not really sure what needs to happen here. Can we set toolstate to "broken" for the RLS? Do we have to (a) open a PR fixing rustfmt, (b) redirect RLS in its cargo.toml to use that PR, then (c) land this patch, then (d) land that PR, then (e) redirect RLS back?

cc @nrc @rust-lang/dev-tools @rust-lang/compiler -- sorry for the broad cc here but I'm actually unclear on who precisely I *ought* to cc!

**Comment by alexcrichton** [comment]
Sure yeah I can write out how this works, sorry we don't have some great instructions for this yet! If it works out here, though, maybe we can write them down in `CONTRIBUTING.md` :)

1. First, create a `config.toml` if it doesn't already exists, set `submodules = false`
2. Run `./x.py test src/tools/rustfmt`, fix any errors in the submodule itself until it works
3. Run `./x.py test src/tools/rls`, fix any errors in the submodule itself until it works
4. Make a commit for `rustfmt`, if necessary, and send a PR to the master branch of rust-lang-nursery/rustfmt
5. Do the same, if necessary for the RLS
6. A maintainer of rls/rustfmt (currently @nrc has done this but others can too!) will **not** merge the PR. The PR can't be merged because CI will be broken. Instead a new branch will be created, and the PR will be pushed to the branch (the PR is left open)
7. On your branch, update the rls/rustfmt submodules to these branches
8. Commit the changes, update your PR to rust-lang/rust
9. Wait for the branch to merge
10. Wait for a nightly
11. A maintainer of rls/rustfmt will merge the original PRs to rls/rustfmt
12. Eventually the rls/rustfmt submodules will get re-updated back to the master branch

Eventually we'll be able to set the tools to "broken" temporarily to skip a lot of these steps, but unfortunately we're not there yet for the rls/rustfmt. :(

Lemme know if there's any questions though!

**Comment by shepmaster** [comment]
@sunjay, it's been over 7 days since we last heard from you. Will you be able to address the build failure [as suggested above](https://github.com/rust-lang/rust/pull/44766#issuecomment-332653525)?

**Comment by sunjay** [comment]
@shepmaster Hi, sorry about that. I've been caught up in some work stuff. I'll be finishing it sometime today or tomorrow.

**Comment by sunjay** [comment]
@alexcrichton I think I've followed your instructions exactly. I just pushed the updated submodules to this PR. 

One thing to note is that the submodules were very out of date so my [PR to rustfmt](https://github.com/rust-lang-nursery/rustfmt/pull/2043) isn't even mergeable. Do you want me to update the submodules to the latest rustfmt master and then fix everything?

In addition to fixing this, I also went ahead and created a PR for adding these instructions to the [CONTRIBUTING.md file](https://github.com/rust-lang/rust/pull/45098). I'll be updating those as I go with any more details that come up as we deal with all of this.

**Comment by sunjay** [comment]
@alexcrichton After talking with @nikomatsakis, we decided it is probably best to try and update rustfmt to the latest master so that we don't have to re-do this later. I was trying to update my rustfmt submodule to the latest master, but when I build, I get the following error message:

```
error: failed to resolve patches for `https://github.com/rust-lang-nursery/rustfmt`

Caused by:
  patch for `rustfmt-nightly` in `https://github.com/rust-lang-nursery/rustfmt` did not resolve to any crates
failed to run: /home/sunjay/Documents/projects/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /home/sunjay/Documents/projects/rust/src/bootstrap/Cargo.toml
```

Do you know what I need to do?

**Comment by alexcrichton** [comment]
@sunjay sorry I'm not super familiar with rustfmt development, the repo I believe has a few branches and a few names, I'm not sure how it all matches up though

**Comment by sunjay** [comment]
@alexcrichton Do you know who I can ask? Do you think it's a problem with the branch? Should I use something other than master?

**Comment by alexcrichton** [comment]
@nrc I believe would know for sure, but he's on leave right now I believe :(

**Comment by sunjay** [comment]
@alexcrichton I believe I was able to figure it out! I made [a PR in cargo](https://github.com/rust-lang/cargo/pull/4607) to improve the error message.

The gist of the problem is that by updating the submodule, the version in rustfmt's `Cargo.toml` changed. This conflicted with the version listed in rustc's `Cargo.lock`. I ran `cargo update -p rustfmt-nightly` to update `Cargo.lock` and now it seems to be building fine.

**Comment by alexcrichton** [comment]
@sunjay holy cow thanks so much for sending a PR to cargo to fix that!

**Comment by sunjay** [comment]
@nikomatsakis @alexcrichton In order to fix rustfmt so this PR is mergeable, I updated to the latest rustfmt. Unfortunately that broke things in other ways and now I'm dependent on [this PR](https://github.com/rust-lang/rust/pull/45096) to be merged before I can continue fixing rustfmt. That PR fixes a bunch of things in src/Cargo.toml which changed when rls changed some stuff.

I can continue with other stuff with regards to generic associated types for now, **but this PR is effectively blocked until https://github.com/rust-lang/rust/pull/45096 is merged** and I make the necessary changes to rustfmt to account for my PR. Maybe I should rebase my branch to be based on that PR's branch?

**Comment by nikomatsakis** [comment]
> Maybe I should rebase my branch to be based on that PR's branch?

Seems like the way to go.

**Comment by shepmaster** [comment]
@nikomatsakis looks like this has been rebased and is ready for re-review!

**Comment by sunjay** [comment]
@nikomatsakis I looked into it further and realized that if I just updated the rustfmt submodule and left rls alone, I could fix everything necessary without needing to be reliant on that other PR I mentioned before. I went ahead and did that and the tests pass locally. Everything is rebased from rust's master branch, so this PR should now be good to go. :)

**Comment by sunjay** [comment]
@alexcrichton @nikomatsakis Looks like clippy is now being built too. My changes break some clippy stuff. Do I need to fix that repo as well before this can be merged? I started the work to do that [here](https://github.com/rust-lang-nursery/rust-clippy/pull/2140).

**Comment by alexcrichton** [comment]
@sunjay can the update to Cargo be avoided here temporarily? It looks like that'll need some extra work elsewhere. Additionally you should be able to declare clippy as "broken" which I believe should allow it to be fixed elsewhere.

**Comment by Manishearth** [comment]
If you are going to declare clippy as broken it would still be nice if you could fix the linting API on this side so we don't have to go back and forth multiple times to get clippy fixed.

**Comment by sunjay** [comment]
@alexcrichton I think I managed to update everything so that cargo doesn't also get updated. Do you still want me to set the toolstate for clippy to broken? I think the build is still failing. Is that because of clippy?

**Comment by alexcrichton** [comment]
Now that we've got support for it, want to just flag clippy/rustfmt/rls to broken and we can land this?

It looks like the clippy submodule is failing to update as a remote ref has gone away? (I don't know the state of all the various PRs related to this one)

**Comment by Manishearth** [comment]
Yeah, sunjay needs to update the ref to the latest on the PR, and then get us to tack that onto a branch

**Comment by sunjay** [comment]
> Now that we've got support for it, want to just flag clippy/rustfmt/rls to broken and we can land this?
> 
> It looks like the clippy submodule is failing to update as a remote ref has gone away? (I don't know the state of all the various PRs related to this one)

Sounds good! That's much easier. :smile: As for the PRs, only rustfmt and clippy are broken as far as I know. (RLS is broken *because* of rustfmt.) The rustfmt PR is already done and is just waiting for review. It can be merged once this lands and the CI passes. The clippy PR will be done by the time this gets merged.

Now that I've cleared this PR and marked the tools as broken, it should be ready for final review and approval from @nikomatsakis. :)

**Comment by sunjay** [comment]
@alexcrichton Any idea why the build is failing?

**Comment by nikomatsakis** [comment]
Looks like [an incremental test is failing](https://travis-ci.org/rust-lang/rust/builds/289314424#L2332).

**Comment by kennytm** [comment]
@bors retry 

`cross` job errored, failed to upload to S3 due to outdated aws-sdk-resources gem.

<details>

```
ERROR: Could not find a valid gem 'aws-sdk-resources' (= 2.10.70) in any repository
ERROR: Possible alternatives: aws-sdk-resources
/home/travis/.rvm/rubies/ruby-2.2.7/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- aws-sdk (LoadError)
	from /home/travis/.rvm/rubies/ruby-2.2.7/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:85:in `requires'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:6:in `<class:S3>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:5:in `<class:Provider>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:4:in `<module:DPL>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider/s3.rb:3:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:59:in `const_get'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:59:in `block in new'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:41:in `fold'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/provider.rb:56:in `new'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:31:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/lib/dpl/cli.rb:7:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.43/bin/dpl:5:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `load'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `<main>'
```

</details>

**Pull request merged**
